packages/test/test-utils/src/loaderContainerTracker.ts

```diff
@@ -184,20 +186,40 @@ export class LoaderContainerTracker implements IOpProcessingController {
                         // Only write it out once
                         waitingSequenceNumberSynchronized = true;
                         debugWait("Waiting for sequence number synchronized");
-                        await this.waitForAnyInboundOps(containersToApply);
+                        await timeoutAwait(this.waitForAnyInboundOps(containersToApply), {
+                            durationMs: timeoutDuration - (Date.now() - start),
+                            errorMsg: "Timeout on waiting for sequence number synchronized",
+                        });
                     }
                 } else {
                     waitingSequenceNumberSynchronized = false;
-                    await this.waitForPendingClients(pendingClients);
+                    await timeoutAwait(this.waitForPendingClients(pendingClients), {
+                        durationMs: timeoutDuration - (Date.now() - start),
+                        errorMsg: "Timeout on waiting for pending join or leave op",
+                    });
                 }
             } else {
                 // Wait for all the containers to be saved
                 // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                 debugWait(`Waiting container to be saved ${dirtyContainers.map((c) => this.containers.get(c)!.index)}`);
+                const remainedDuration = timeoutDuration - (Date.now() - start);
                 waitingSequenceNumberSynchronized = false;
                 await Promise.all(dirtyContainers.map(async (c) => Promise.race(
-                    [new Promise((resolve) => c.once("saved", resolve)),
-                    new Promise((resolve) => c.once("closed", resolve))],
+                    [timeoutPromise(
+                        (resolve) => c.once("saved", () => resolve()),
+                        {
+                            durationMs: remainedDuration / dirtyContainers.length,
```

Comment:
This part needs waiting all dirty containers to be closed or saved. So I allocate the remained duration time average to each container. @anthony-murphy  I am afraid if each duration is too short in the real test cases, but as we discussed user can define their own timeout durations for different tests.

Reply:
we don't need to divide. all these waits happen in parallel, so they should all get the full amout of time